include/core/Discretization.hpp

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+struct Discretization
+{
+    virtual ~Discretization() = default;
+
+    enum GridType { hydrofabric_locations, spatial_structured, spatial_unstructured };
+    virtual GridType grid_type() = 0;
+
+    // Must cover an identified geographic domain
+    virtual GeoDivide divide() = 0;
+
+    virtual size_t size() = 0;
+    virtual int rank() = 0;
+    virtual std::span<Foo> get_points() = 0;
```

Comment:
I think technically, we may want to distinguish between a discretization over `Catchment`, one over a designated `FlowPath` within said `Catchment`, and the outflow `Nexus` that those feed. Maybe even template `Discretization` over those possibilities as well as `GeometricPoint`, so that `get_points` can return the right kind of collection, rather than numbers with arbitrarily different meanings.

If we distinguished a `Discretization<Catchment>` from a matching `Discretization<Nexus>`, we'd need to be a bit more sophisticated in expressing the 'simple' one-to-one coupling cases, but I think that may be worthwhile. Additionally, if the number or ordering of those actually differs in the input data, then we can't just map those one-to-one anyway.